pkg/diff/chunk.go

```diff
@@ -17,260 +17,530 @@ import (
-		max, ok2 := chunk.end.(int64)
-		if !ok1 || !ok2 {
-			return nil, errors.Errorf("can't parse chunk's begin: %v, end: %v", chunk.begin, chunk.end)
-		}
-		step := (max - min + count - 1) / count
-		cutoff := min
-		for cutoff <= max {
-			r := newChunkRange(cutoff, cutoff+step, true, false, false, false)
-			chunks = append(chunks, r)
-			cutoff += step
-		}
+	return s.getChunksByBuckets()
+}
+
+func (s *bucketSpliter) getChunksByBuckets() ([]*chunkRange, error) {
+	chunks := make([]*chunkRange, 0, 1000)
 
-		log.Debugf("getChunksForTable cut table: cnt=%d min=%v max=%v step=%v chunk=%d", count, min, max, step, len(chunks))
-	} else if reflect.TypeOf(chunk.begin).Kind() == reflect.Float64 {
-		min, ok1 := chunk.begin.(float64)
-		max, ok2 := chunk.end.(float64)
-		if !ok1 || !ok2 {
-			return nil, errors.Errorf("can't parse chunk's begin: %v, end: %v", chunk.begin, chunk.end)
+	indices := dbutil.FindAllIndex(s.table.info)
+	for _, index := range indices {
+		if index == nil {
+			continue
 		}
-		step := (max - min + float64(count-1)) / float64(count)
-		cutoff := min
-		for cutoff <= max {
-			r := newChunkRange(cutoff, cutoff+step, true, false, false, false)
-			chunks = append(chunks, r)
-			cutoff += step
+		buckets, ok := s.buckets[index.Name.O]
+		if !ok {
+			return nil, errors.NotFoundf("index %s in buckets info", index.Name.O)
 		}
 
-		log.Debugf("getChunksForTable cut table: cnt=%d min=%v max=%v step=%v chunk=%d",
-			count, min, max, step, len(chunks))
-	} else {
-		max, ok1 := chunk.end.(string)
-		min, ok2 := chunk.begin.(string)
-		if !ok1 || !ok2 {
-			return nil, errors.Errorf("can't parse chunk's begin: %v, end: %v", chunk.begin, chunk.end)
-		}
+		var (
+			chunkNum    int64
+			lowerValues []string
+			upperValues []string
+			latestCount int64
+			err         error
+		)
 
-		// get random value as split value
-		splitValues, err := dbutil.GetRandomValues(context.Background(), db, Schema, table, column.Name.O, count-1, min, max, limitRange, collation)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
+		indexColumns := getColumnsFromIndex(index, s.table.info)
 
-		var minTmp, maxTmp string
-		var i int64
-		for i = 0; i < int64(len(splitValues)+1); i++ {
-			if i == 0 {
-				minTmp = min
-			} else {
-				minTmp = fmt.Sprintf("%s", splitValues[i-1])
+		for i, bucket := range buckets {
+			upperValues, err = dbutil.AnalyzeValuesFromBuckets(bucket.UpperBound, indexColumns)
+			if err != nil {
+				return nil, errors.Trace(err)
 			}
-			if i == int64(len(splitValues)) {
-				maxTmp = max
-			} else {
-				maxTmp = fmt.Sprintf("%s", splitValues[i])
+
+			if bucket.Count-latestCount > int64(s.chunkSize) || i == len(buckets)-1 {
+				// create a new chunk
+				chunk := newChunkRange()
+				var lower, upper, lowerSymbol, upperSymbol string
+				for j, col := range index.Columns {
+					if len(lowerValues) != 0 {
+						lower = lowerValues[j]
+						lowerSymbol = gt
+					}
+					if i != len(buckets)-1 {
```

Comment:
`if i != len(buckets)-1 {`, do you confirm this. I can not understand again. :anguished:

Reply:
this mean if the bucket is the last one, we should not set the upper, for example the last bucket in target is [10, 11], it may exist data greater than 11 in source, so just generate chunk [10, ) is ok.